clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -972,16 +975,28 @@ private Record nextFetchedRecord() {
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
             if (isFetched)
                 return Collections.emptyList();
+            if (nextInlineException != null) {
```

Comment:
Spoke to @hachikuji offline. There are two things to be resolved:
1. Ensure that the good records before a bad records in the same batch is delivered.
2. Do not skip over a bad record.
This patch does 1 but not 2. I have created KAFKA-5211 to address 2 and will submit a patch for that separately. For this patch we will just focus on solving 1.